app/controllers/registrations_controller.rb

```diff
@@ -378,6 +382,53 @@ def stripe_webhook
 
       stored_intent = stored_record.payment_intent
       stored_intent.update_status_and_charges(connected_account, stripe_intent, audit_event, audit_remote_timestamp)
+    when StripeWebhookEvent::REFUND_CREATED, StripeWebhookEvent::REFUND_UPDATED
+      # stripe_intent contains a Stripe::Refund as per Stripe documentation
+
+      original_charge = StripeRecord.charge.find_by(stripe_id: stripe_intent.charge)
+
+      if original_charge.nil?
+        # We created a record because refund creation is an incoming event,
+        #   but now we figured out that we don't care about it after all.
+        stored_record.destroy
+
+        logger.error "Stripe webhook reported a refund on charge #{stripe_intent.charge} but we never issued that one"
+        return head :not_found
+      elsif stored_record.parent_record.nil?
+        stored_record.update!(parent_record: original_charge)
+      end
+
+      # NOTE: Feels weird to be checking the type of the event in this block - is there a better way to handle this logic?
+      stored_record.update_from_api(stripe_intent) if event.type == StripeWebhookEvent::REFUND_UPDATED
+      stored_intent = stored_record.root_record.payment_intent
+      stored_holder = stored_intent.holder
+
+      if stored_holder.is_a? Registration
+        ruby_money = stored_record.money_amount
+
+        original_payment = RegistrationPayment.find_by(receipt: original_charge)
+
+        if original_payment.nil?
+          logger.error "Tried to record refund on charge #{stripe_intent.charge} but we never issued that one"
+          return head :not_found
+        end
+
+        stored_holder.with_lock do
+          already_refunded = stored_record.registration_payment.present?
+
+          refund_succeeded = stored_record.stripe_status_succeeded?
```

Comment:
This check specifically can happen outside of the lock

Reply:
Fixed